File: demos/bird_tracker.ex
defmodule BirdTracker do
  @type sighting :: %{species: String.t(), count: pos_integer(), habitat: atom()}

  def summarize(sightings) do
    Enum.reduce(sightings, %{}, fn %{habitat: habitat, count: count}, totals ->
      Map.update(totals, habitat, count, &(&1 + count))
    end)
  end

  def label(%{species: species, count: count, habitat: habitat}) do
    "#{count} #{species} in the #{habitat}"
  end
end

BirdTracker.summarize([%{species: "Bullfinch", count: 2, habitat: :garden}])

File: demos/bird_tracker.dart
enum Habitat { garden, woodland, wetland }

class Sighting {
  const Sighting(this.species, this.count, this.habitat);

  final String species;
  final int count;
  final Habitat habitat;

  String get label => '$count $species in ${habitat.name}';
}

Map<Habitat, int> summarize(Iterable<Sighting> sightings) {
  final totals = <Habitat, int>{};
  for (final sighting in sightings) {
    totals.update(sighting.habitat, (count) => count + sighting.count, ifAbsent: () => sighting.count);
  }
  return totals;
}

void main() {
  final sightings = [const Sighting('Bullfinch', 2, Habitat.garden)];
  print(summarize(sightings));
}
